### ISSUE TITLE
Nextcloud clients send corrupted files than fixes it

### ISSUE THREAD

**Issue by dbeniamine** [nextcloud/desktop#2571]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Expected behaviour

Files that are not modified should not be synchronized and files that are not locally corrupted should not be corrupted on the server

### Actual behaviour

This bug has happened twice to me on the last days:

1. The nextcloud client "decides" that some files (that have not been touched) should be uploaded
2. Some of the files send to the server are corrupted
3. The nextcloud clients re-upload the files fixing the corruption issue

In the end all is well as the clients fixes the issue without human intervention, but the client has sync several time a few giga bytes of data and during the process the files where corrupted on the server and on all synced computers but mine.

There are two problems here:

1. Files where re-upload even if I did not touch it (did not even open them)
2. Some files where corrupted during the process

### Steps to reproduce

1. Upgrade from 2.7 to 3.0.1 maybe ?
2. Wait until it happens

### Client configuration
Client version: Version 3.0.1 (Debian).

Operating system: Debian testing

OS language: English

Qt version used by client package (Linux only, see also Settings dialog): 
QT 5.14.2

Client package (From Nextcloud or distro) (Linux only):

nextcloud-desktop                        3.0.1-3 from debian

Installation path of client:

/usr/bin/nextcloud

### Server configuration

Nextcloud version: 19.0.3

Storage backend (external storage): None

### Logs

Please use Gist (https://gist.github.com/) or a similar code paster for longer
logs.

1. Client logfile: 

I do not have the logs of the moments when Nextcloud client start syncing old files

But on the server side if I look at one file I have the following activity, showing modifications twis week:

![image](https://user-images.githubusercontent.com/2989133/96604683-ac2f4a80-12f5-11eb-8841-2b3a03dca774.png)

But only one version (3 years ago) : 

![image](https://user-images.githubusercontent.com/2989133/96604758-c0734780-12f5-11eb-9a46-ccac32147b19.png)

And current version is also 3 year old:

![image](https://user-images.githubusercontent.com/2989133/96605050-0c25f100-12f6-11eb-91f6-41348709c488.png)

On my filesystem this is a result of stat:

```  
File: /home/david/.../PR1712-0027.pdf
  Size: 48458     	Blocks: 112        IO Block: 4096   regular file
Device: 36h/54d	Inode: 10359202    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   david)   Gid: ( 1000/   david)
Access: 2020-10-19 23:10:26.379362959 +0200
Modify: 2017-12-15 13:11:58.000000000 +0100
Change: 2019-05-13 19:56:42.355998960 +0200
 Birth: -
```

2. Web server error log:

3. Server logfile: nextcloud log (data/nextcloud.log):

```
[webdav] Fatal: Sabre\DAV\Exception\BadRequest: Expected filesize of 1163264 bytes but read (from Nextcloud client) and wrote (to Nextcloud storage) 1048576 bytes. Could either be a network problem on the sending side or a problem writing to the storage on the server side. at <<closure>>

0. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 1143
   OCA\DAV\Connector\Sabre\File->put(null)
1. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 515
   Sabre\DAV\Server->updateFile("*** sensitive parameters replaced ***")
2. /var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php line 89
   Sabre\DAV\CorePlugin->httpPut(Sabre\HTTP\Request {}, Sabre\HTTP\Response {})
3. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 474
   Sabre\DAV\Server->emit("method:PUT", [Sabre\HTTP\Requ ... }])
4. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 251
   Sabre\DAV\Server->invokeMethod(Sabre\HTTP\Request {}, Sabre\HTTP\Response {})
5. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 319
   Sabre\DAV\Server->start()
6. /var/www/nextcloud/apps/dav/lib/Server.php line 320
   Sabre\DAV\Server->exec()
7. /var/www/nextcloud/apps/dav/appinfo/v2/remote.php line 35
   OCA\DAV\Server->exec()
8. /var/www/nextcloud/remote.php line 167
   require_once("/var/www/nextcl ... p")

PUT /remote.php/dav/files/david/<Path removed>
from <IP Removed> by david at 2020-10-18T15:31:50+00:00
```

**Comment by er-vin** [comment]
This is going to be tough to locate since it seems to happen randomly... I'd say you'd need to turn on debug logs to log constantly with a long expire (this will be default on 3.1) and when it happens you should save the logs somewhere and give us access to them letting us now which file was wrongly uploaded.

That's the only way for us to know a) why it decided to upload and b) how the upload went.

**Comment by dbeniamine** [comment]
> This is going to be tough to locate since it seems to happen randomly... I'd say you'd need to turn on debug logs to log constantly with a long expire (this will be default on 3.1)

I agree with that, but how do I set it ?

Should I always launch the client with something like `nextcloud --logdebug --logdir /home/tmp/nextcloud.log`

**Comment by camilasan** [comment]
> Should I always launch the client with something like `nextcloud --logdebug --logdir /home/tmp/nextcloud.log`

yes, but please correct the argument for --logdir to ```/home/tmp``` (it is just the folder)

thanks!

**Comment by er-vin** [comment]
Also note that for clients 3.0 and later, you can do it via config, typically by adding the following in the nextcloud.cfg General group:
```
logDebug=true
logDir=/home/tmp/logs
logExpire=24
```

I strongly advise the logExpire if you log constantly like this, because otherwise you might fill up your disks fast. With that option the logs will be purged after 24h.

**Comment by dbeniamine** [comment]
I am closing this issue, I believe that (but am not sure) the problem was not due to Nextcloud but ecryptfs.

I had an unstandard ecryptfs configuration and independently of nextcloud I found that sometimes some of my files were not correctly decrypted. I managed to decrypt them all, re-do my install with a full disk luks encryption, since that the issue never came back.

So maybe nextcloud was reacting to this decryption fails uploading crypted files and then reuploading the decrypted files once they wee avaiable again.
